### PR TITLE
Record downtime in ac-down attribute

### LINKED ISSUE #332: zino1 protocol: rechability events is missing the ac-down parameter
It looks like the reachability(and portstates? haven't checked) events are missing the `ac-down` parameter.
This parameter is a shorthand for `accumulated downtime` for a event and counts seconds of downtime for the port/device for the life of that event.

### PR DESCRIPTION
**Pull request by stveit**
## Scope and purpose

Fixes #332. 

Makes it so downtime is recorded for events when they are committed. This requires changes to the event to be done via the
Checkout -> commit workflow. A change had to be done in `reachabletask.py`, where previously it just fetched an event and made changes to the event object directly instead of checking it out.

This is not a direct port from Zino1, but the behaviour should be the same,

## Contributor Checklist

Every pull request should have this checklist filled out, no matter how small it is.
More information about contributing to Zino can be found in the
[README](https://github.com/Uninett/zino/blob/master/README.md#developing-zino).

* [x] Added a changelog fragment for [towncrier](https://github.com/Uninett/zino/blob/master/README.md#before-merging-a-pull-request)
* [x] Added/amended tests for new/changed code 
* [ ] Added/changed documentation
* [x] Linted/formatted the code with black, ruff and isort, easiest by using [pre-commit](https://github.com/Uninett/zino/blob/master/README.md#code-style)
* [x] The first line of the commit message continues the sentence "If applied, this commit will ...", starts with a capital letter, does not end with punctuation and is 50 characters or less long. See https://cbea.ms/git-commit/
* [ ] If applicable: Created new issues if this PR does not fix the issue completely/there is further work to be done

### CODE REVIEW

File: src/zino/tasks/reachabletask.py [hunk 1]
```diff
@@ -50,8 +50,10 @@ async def _run_extra_job(self):
     def _update_reachability_event_as_reachable(self):
         event = self.state.events.get(self.device.name, None, ReachabilityEvent)
         if event and event.reachability != ReachabilityState.REACHABLE:
-            event.reachability = ReachabilityState.REACHABLE
+            event = self.state.events.checkout(event.id)
             event.add_log(f"{self.device.name} reachable")
+            event.reachability = ReachabilityState.REACHABLE
+            self.state.events.commit(event)
```
Comment (lunkwill42) [inline src/zino/tasks/reachabletask.py:56]: This looks like an unrelated bugfix that should be in a PR of its own, or?

File: tests/events_test.py [hunk 2]
```diff
@@ -255,3 +255,41 @@ def test_delete_should_remove_closed_event_from_index_if_still_in_index(self):
 
         index = EventIndex("foobar", None, ReachabilityEvent)
         assert not events._events_by_index.get(index)
+
+    def test_record_downtime_should_not_update_event_if_lasttrans_is_not_set(self):
```
Comment (lunkwill42) [inline tests/events_test.py:259]:
```suggestion
    def test_when_lasttrans_is_not_set_record_downtime_should_not_update_event(self):
```

File: tests/events_test.py [hunk 3]
```diff
@@ -255,3 +255,41 @@ def test_delete_should_remove_closed_event_from_index_if_still_in_index(self):
...
+        new_event.reachability = ReachabilityState.REACHABLE
+        new_event.lasttrans = None
+        new_event.ac_down = None
+
+        events.record_downtime(new_event, old_event)
+
+        assert new_event.lasttrans is None
+        assert new_event.ac_down is None
+
+    def test_record_downtime_should_not_update_event_if_downtime_is_calculated_to_zero_or_less(self, monkeypatch):
```
Comment (lunkwill42) [inline tests/events_test.py:275]:
```suggestion
    def test_when_downtime_is_calculated_to_zero_or_less_record_downtime_should_not_update_event(self, monkeypatch):
```

File: tests/tasks/test_linkstatetask.py [hunk 4]
```diff
@@ -77,6 +79,97 @@ def test_when_interface_data_is_empty_update_interfaces_should_keep_processing(s
             is None
         )
 
+    @pytest.mark.asyncio
+    async def test_should_set_lasttrans_for_new_portstate_event(self, linkstatetask_with_one_link_down):
```
Comment (lunkwill42) [inline tests/tasks/test_linkstatetask.py:83]:
```suggestion
    async def test_when_event_is_new_it_should_set_lasttrans(self, linkstatetask_with_one_link_down):
```

File: tests/tasks/test_linkstatetask.py [hunk 5]
```diff
@@ -77,6 +79,97 @@ def test_when_interface_data_is_empty_update_interfaces_should_keep_processing(s
...
+        task = linkstatetask_with_one_link_down
+        await task.run()
+        event = task.state.events.get(task.device.name, 2, PortStateEvent)
+        assert event.lasttrans
+
+    @pytest.mark.asyncio
+    async def test_should_update_lasttrans_for_portstate_event_going_from_down_to_up(self, linkstatetask_with_links_up):
```
Comment (lunkwill42) [inline tests/tasks/test_linkstatetask.py:90]:
```suggestion
    async def test_when_event_transitions_from_to_up_it_should_update_lasttrans(self, linkstatetask_with_links_up):
```

File: tests/tasks/test_linkstatetask.py [hunk 6]
```diff
@@ -77,6 +79,97 @@ def test_when_interface_data_is_empty_update_interfaces_should_keep_processing(s
...
+        event.lasttrans = initial_lasttrans
+        task.state.events.commit(event)
+
+        assert (await task.run()) is None
+        updated_event = task.state.events[event.id]
+        assert updated_event.portstate == InterfaceState.UP
+        assert updated_event.lasttrans > initial_lasttrans
+
+    @pytest.mark.asyncio
+    async def test_should_update_lasttrans_for_portstate_event_going_from_down_to_admindown(
```
Comment (lunkwill42) [inline tests/tasks/test_linkstatetask.py:109]:
```suggestion
    async def test_when_event_transitions_from_down_to_admindown_it_should_update_lasttrans(
```

File: tests/tasks/test_linkstatetask.py [hunk 7]
```diff
@@ -77,6 +79,97 @@ def test_when_interface_data_is_empty_update_interfaces_should_keep_processing(s
...
+        event.lasttrans = initial_lasttrans
+        task.state.events.commit(event)
+
+        assert (await task.run()) is None
+        updated_event = task.state.events[event.id]
+        assert updated_event.portstate == InterfaceState.ADMIN_DOWN
+        assert updated_event.lasttrans > initial_lasttrans
+
+    @pytest.mark.asyncio
+    async def test_should_update_ac_down_for_portstate_event_going_from_down_to_up(self, linkstatetask_with_links_up):
```
Comment (lunkwill42) [inline tests/tasks/test_linkstatetask.py:130]:
```suggestion
    async def test_when_event_transitions_from_down_to_up_it_should_update_ac_down(self, linkstatetask_with_links_up):
```

File: tests/tasks/test_linkstatetask.py [hunk 8]
```diff
@@ -77,6 +79,97 @@ def test_when_interface_data_is_empty_update_interfaces_should_keep_processing(s
...
+        event.lasttrans = initial_lasttrans
+        task.state.events.commit(event)
+
+        assert (await task.run()) is None
+        updated_event = task.state.events[event.id]
+        assert updated_event.portstate == InterfaceState.UP
+        assert updated_event.ac_down > initial_ac_down
+
+    @pytest.mark.asyncio
+    async def test_should_update_ac_down_for_portstate_event_going_from_down_to_admin_down(
```
Comment (lunkwill42) [inline tests/tasks/test_linkstatetask.py:151]:
```suggestion
    async def test_when_event_transitions_from_down_to_admindown_it_should_update_ac_down(
```

File: tests/tasks/test_reachabletask.py [hunk 9]
```diff
@@ -66,3 +71,63 @@ async def test_run_extra_job_should_not_update_event_when_device_is_unreachable(
         event.reachability = ReachabilityState.NORESPONSE
         assert (await task._run_extra_job()) is None
         assert event.reachability == ReachabilityState.NORESPONSE
+
+    @pytest.mark.asyncio
+    async def test_should_set_lasttrans_for_new_reachable_event(self, unreachable_task):
```
Comment (lunkwill42) [inline tests/tasks/test_reachabletask.py:76]:
```suggestion
    async def test_when_event_is_new_it_should_set_lasttrans(self, unreachable_task):
```

File: tests/tasks/test_reachabletask.py [hunk 10]
```diff
@@ -66,3 +71,63 @@ async def test_run_extra_job_should_not_update_event_when_device_is_unreachable(
...
+        task = unreachable_task
+        with pytest.raises(DeviceUnreachableError):
+            await task.run()
+        event = task.state.events.get(task.device.name, None, ReachabilityEvent)
+        assert event.lasttrans
+
+    @pytest.mark.asyncio
+    async def test_should_update_lasttrans_for_reachable_event_going_from_no_response_to_reachable(
```
Comment (lunkwill42) [inline tests/tasks/test_reachabletask.py:84]:
```suggestion
    async def test_when_event_transitions_from_no_response_to_reachable_it_should_update_lasttrans(
```

File: tests/tasks/test_reachabletask.py [hunk 11]
```diff
@@ -66,3 +71,63 @@ async def test_run_extra_job_should_not_update_event_when_device_is_unreachable(
...
+        event.lasttrans = initial_lasttrans
+        task.state.events.commit(event)
+
+        assert (await task.run()) is None
+        updated_event = task.state.events[event.id]
+        assert updated_event.reachability == ReachabilityState.REACHABLE
+        assert updated_event.lasttrans > initial_lasttrans
+
+    @pytest.mark.asyncio
+    async def test_should_update_lasttrans_for_reachable_event_going_from_reachable_to_no_response(
```
Comment (lunkwill42) [inline tests/tasks/test_reachabletask.py:101]:
```suggestion
    async def test_when_event_transitions_from_reachable_to_no_response_it_should_update_lasttrans(
```

File: tests/tasks/test_reachabletask.py [hunk 12]
```diff
@@ -66,3 +71,63 @@ async def test_run_extra_job_should_not_update_event_when_device_is_unreachable(
...
+        task.state.events.commit(event)
+
+        with pytest.raises(DeviceUnreachableError):
+            await task.run()
+        updated_event = task.state.events[event.id]
+        assert updated_event.reachability == ReachabilityState.NORESPONSE
+        assert updated_event.lasttrans > initial_lasttrans
+
+    @pytest.mark.asyncio
+    async def test_should_update_ac_down_for_reachable_event_going_from_no_response_to_reachable(self, reachable_task):
```
Comment (lunkwill42) [inline tests/tasks/test_reachabletask.py:119]:
```suggestion
    async def test_when_event_transitions_from_no_response_to_reachable_it_should_update_ac_down(self, reachable_task):
```

File: tests/trapobservers/link_traps_test.py [hunk 13]
```diff
@@ -119,6 +119,23 @@ async def test_when_link_trap_refers_to_unknown_port_it_should_ignore_trap_early
             assert not await observer.handle_trap(trap)
             assert not handle_link_transition.called, "handle_link_transition was called"
 
+    async def test_should_set_lasttrans_for_new_portstate_event(
```
Comment (lunkwill42) [inline tests/trapobservers/link_traps_test.py:122]:
```suggestion
    async def test_when_event_is_new_it_should_set_lasttrans(
```

File: tests/trapobservers/link_traps_test.py [hunk 14]
```diff
@@ -244,3 +261,84 @@ def test_when_link_transitions_to_non_flapping_it_should_remove_flap_state(self,
         observer.handle_link_transition(localhost, port, is_up=True)
 
         assert index not in state_with_localhost_with_port.flapping.interfaces
+
+    def test_when_link_transitions_from_up_to_flapping_lasttrans_for_related_event_should_be_updated(
```
Comment (lunkwill42) [inline tests/trapobservers/link_traps_test.py:265]:
```suggestion
    def test_when_link_transitions_from_up_to_flapping_it_should_update_lasttrans_for_related_event(
```

File: tests/trapobservers/link_traps_test.py [hunk 15]
```diff
@@ -244,3 +261,84 @@ def test_when_link_transitions_to_non_flapping_it_should_remove_flap_state(self,
...
+
+        observer.handle_link_transition(localhost, port, is_up=False)
+
+        updated_event = events[event.id]
+
+        assert updated_event.flapstate == FlapState.FLAPPING
+        assert updated_event.portstate == InterfaceState.DOWN
+        assert updated_event.lasttrans > initial_lasttrans
+
+    async def test_when_link_transitions_from_flapping_to_up_lasttrans_for_related_event_should_be_updated(
```
Comment (lunkwill42) [inline tests/trapobservers/link_traps_test.py:290]:
```suggestion
    async def test_when_link_transitions_from_flapping_to_up_it_should_update_lasttrans_for_related_event(
```

File: tests/trapobservers/link_traps_test.py [hunk 16]
```diff
@@ -244,3 +261,84 @@ def test_when_link_transitions_to_non_flapping_it_should_remove_flap_state(self,
...
+        events.commit(event)
+
+        observer.handle_link_transition(localhost, port, is_up=True)
+
+        updated_event = events[event.id]
+        assert updated_event.flapstate == FlapState.STABLE
+        assert updated_event.portstate == InterfaceState.UP
+        assert updated_event.lasttrans > initial_lasttrans
+
+    async def test_when_link_transitions_from_flapping_to_up_ac_down_for_related_event_should_be_updated(
```
Comment (lunkwill42) [inline tests/trapobservers/link_traps_test.py:317]:
```suggestion
    async def test_when_link_transitions_from_flapping_to_up_it_should_update_ac_down_for_related_event(
```

**Review by lunkwill42 — CHANGES_REQUESTED** [review]
You know me, I love picking at test names.

Otherwise, it looks pretty great, just as when we reviewed IRL.

File: src/zino/tasks/reachabletask.py (hunk 1 above)
Comment (stveit) [reply src/zino/tasks/reachabletask.py:56]: I wouldnt call it unrelated since this is a necessary fix for the ac-down stuff to work properly for reachability events, but could be its own PR that this one depends on

Comment (lunkwill42) [reply src/zino/tasks/reachabletask.py:56]: It should normally be - splitting things up always makes the PRs easier to review - but I'm going to let it slide this time, since I've already reviewed what I want to review.  Suggest @johannaengland  look at also :)

**Review by johannaengland — APPROVED** [review]
